go/vt/vttablet/tabletserver/messager/message_manager.go

```diff
@@ -252,7 +252,7 @@ func newMessageManager(tsv TabletService, vs VStreamer, table *schema.Table, pos
 		}},
 	}
 	mm.readByPriorityAndTimeNext = sqlparser.BuildParsedQuery(
-		"select priority, time_next, epoch, time_acked, %s from %v where time_next < %a order by priority, time_next desc limit %a",
+		"select priority, time_next, epoch, time_acked, %s from %v where time_acked is null and time_next < %a order by priority, time_next desc limit %a",
```

Comment:
I'm surprised that this showed a query plan difference, assuming that you have an index on `time_next`

### original

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "1.71"
    },
    "ordering_operation": {
      "using_filesort": true,
      "cost_info": {
        "sort_cost": "1.00"
      },
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "range",
        "possible_keys": [
          "time_next"
        ],
        "key": "time_next",
        "used_key_parts": [
          "time_next"
        ],
        "key_length": "9",
        "rows_examined_per_scan": 1,
        "rows_produced_per_join": 1,
        "filtered": "100.00",
        "index_condition": "(`keywords`.`searches__extractor__msgs`.`time_next` < <cache>((unix_timestamp(now()) * 1e9)))",
        "cost_info": {
          "read_cost": "0.61",
          "eval_cost": "0.10",
          "prefix_cost": "0.71",
          "data_read_per_join": "728"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ]
      }
    }
  }
}
```

### with time_acked is null
```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "1.21"
    },
    "ordering_operation": {
      "using_filesort": true,
      "cost_info": {
        "sort_cost": "0.50"
      },
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "range",
        "possible_keys": [
          "ack_idx",
          "time_next"
        ],
        "key": "time_next",
        "used_key_parts": [
          "time_next"
        ],
        "key_length": "9",
        "rows_examined_per_scan": 1,
        "rows_produced_per_join": 0,
        "filtered": "50.00",
        "index_condition": "(`keywords`.`searches__extractor__msgs`.`time_next` < <cache>((unix_timestamp(now()) * 1e9)))",
        "cost_info": {
          "read_cost": "0.66",
          "eval_cost": "0.05",
          "prefix_cost": "0.71",
          "data_read_per_join": "363"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ],
        "attached_condition": "(`keywords`.`searches__extractor__msgs`.`time_acked` is null)"
      }
    }
  }
}
```

Reply:
@derekperkins Yes, this would only serve to avoid sending rows up the stack that will only be ignored when we go to add it to the message cache w/o some table side changes. I added a comment about what index I think should be there. The key factors are:
1. I can't imaging why the poller would ever need to read already acked messages. I would assume that this would often filter out a large portion of the rows and at a minimum it prevents us from returning rows to the message manager that we know won't end up added to the cache in the end
2. An *new* index on `(time_acked, priority, time_next desc)` should allow us to:
    - Push the index conditions down to the storage engine ([index condition pushdown](https://dev.mysql.com/doc/refman/8.0/en/index-condition-pushdown-optimization.html)) — which you'll see identified in the standard `EXPLAIN` output with `Using index condition` and JSON output with `"index_condition:" ...` — so that the filtering can be done there w/o having to pass information back through the storage engine API to the mysql query execution layer
    -  Avoid having to do a filesort — which we will see in the standard `EXPLAIN` output when `Using filesort` is absent and in the JSON output when we see `"using_filesort": false` — as we can get the order  from the index records that match `time_acked` predicate **_and_** apply the limit in the storage engine as well

So the new suggested/documented message table structure would be (also changing the `time_next` default to -1 so that messages are not skipped/lost when the binlog streamer connection is restarted):
```
create table my_message(
  time_scheduled bigint,
  id bigint,
  time_next bigint DEFAULT -1,
  epoch bigint,
  time_created bigint,
  time_acked bigint,
  message varchar(128),
  priority tinyint NOT NULL DEFAULT '0',
  primary key(time_scheduled, id),
  unique index id_idx(id),
  index poller_idx (time_acked, priority asc, time_next desc)
) comment 'vitess_message,vt_ack_wait=30,vt_purge_after=86400,vt_batch_size=10,vt_cache_size=10000,vt_poller_interval=30'
```
    
This should allow us to do the minimum amount of work necessary up the stack: InnoDB->MySQL->MessageManager->heap/cache. For example:
```
mysql> select count(*) from mt where time_acked is null and time_next < UNIX_TIMESTAMP(NOW())*1e9;
+----------+
| count(*) |
+----------+
|      463 |
+----------+
1 row in set (0.00 sec)

mysql> explain format=JSON select priority, time_next, epoch, time_acked from mt where time_acked is null and time_next < UNIX_TIMESTAMP(NOW())*1e9 order by priority, time_next desc limit 10000\G
*************************** 1. row ***************************
EXPLAIN: {
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "57.87"
    },
    "ordering_operation": {
      "using_filesort": false,
      "table": {
        "table_name": "mt",
        "access_type": "ref",
        "possible_keys": [
          "poller_idx",
          "poller_idx2"
        ],
        "key": "poller_idx",
        "used_key_parts": [
          "time_acked"
        ],
        "key_length": "9",
        "ref": [
          "const"
        ],
        "rows_examined_per_scan": 463,
        "rows_produced_per_join": 154,
        "filtered": "33.33",
        "index_condition": "((`vt_commerce`.`mt`.`time_acked` is null) and (`vt_commerce`.`mt`.`time_next` < <cache>((unix_timestamp(now()) * 1e9))))",
        "cost_info": {
          "read_cost": "11.57",
          "eval_cost": "15.43",
          "prefix_cost": "57.88",
          "data_read_per_join": "85K"
        },
        "used_columns": [
          "time_scheduled",
          "id",
          "time_next",
          "epoch",
          "time_acked",
          "priority"
        ]
      }
    }
  }
}
1 row in set, 1 warning (0.00 sec)
```

Would you be interested in testing?

Reply:
The new index doesn't seem to have changed anything

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "1.21"
    },
    "ordering_operation": {
      "using_filesort": true,
      "cost_info": {
        "sort_cost": "0.50"
      },
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "range",
        "possible_keys": [
          "ack_idx",
          "time_next",
          "time_acked__priority__time_next"
        ],
        "key": "time_next",
        "used_key_parts": [
          "time_next"
        ],
        "key_length": "9",
        "rows_examined_per_scan": 1,
        "rows_produced_per_join": 0,
        "filtered": "50.00",
        "index_condition": "(`keywords`.`searches__extractor__msgs`.`time_next` < <cache>((unix_timestamp(now()) * 1e9)))",
        "cost_info": {
          "read_cost": "0.66",
          "eval_cost": "0.05",
          "prefix_cost": "0.71",
          "data_read_per_join": "364"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ],
        "attached_condition": "(`keywords`.`searches__extractor__msgs`.`time_acked` is null)"
      }
    }
  }
}
```

Reply:
Hi @derekperkins 

Can you please include the query along with the explain plan? Can you also please share the index definitions? From your explain output here, I *think* you're using the new updated/proposed query but the new `time_acked__priority__time_next` index is not in DESC order for the time_next field.

Reply:
Some other changes that haven't made it to the docs yet from v6, some of which change the table structure. https://github.com/vitessio/vitess/releases/tag/v6.0.20-20200429. I suggested some changes in your docs PR

> Messaging: First Class Citizen + API Improvements
> The messaging feature has had several breaking changes and feature additions
> 
> - Direct inserts are now supported: INSERT INTO msg_table SELECT col1, col2...
> - time_scheduled and time_created are no longer required nor automatically populated.
> - priority is a new required field. Messages with a lower priority will be processed first.
> - Message retention is now measured after time_acked instead of time_scheduled
> - New table options vt_min_backoff and vt_max_backoff set bounds on exponential backoff for message retries
> - 33% backoff jitter has been added for messages that are getting postponed to prevent thundering herds.
> 
> Users of the messages feature are encouraged to read through the full details in issue [#5947](https://github.com/vitessio/vitess/issues/5947) [#5948](https://github.com/vitessio/vitess/pull/5948) [#6114](https://github.com/vitessio/vitess/pull/6114)

Reply:
MySQL always chooses the `time_next` index. In a very unscientific test, I ssh'd into the MySQL instance in question to remove all network and Vitess latency, and repeatedly ran queries to get a decent average. The table in question has:
- 9.5M rows
- 6.5M with `time_next=NULL`, `time_acked=12345`
- 3M "failed" with very high  but < 10 matched the query on any given run.

#### Test results:
- Non force index: < 5 ms
- Force index: ~50 ms

```sql
select priority, time_next, epoch, time_acked, id, keyspace_id 
from searches__extractor__msgs # optional force index (time_acked__priority__time_next)
where time_acked is null and time_next < UNIX_TIMESTAMP(NOW())*1e9 
order by priority, time_next desc limit 10000
```

#### Force Index query plan

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "268.10"
    },
    "ordering_operation": {
      "using_filesort": false,
      "table": {
        "table_name": "searches__request_dispatcher__msgs",
        "access_type": "ref",
        "possible_keys": [
          "time_acked__priority__time_next"
        ],
        "key": "time_acked__priority__time_next",
        "used_key_parts": [
          "time_acked"
        ],
        "key_length": "9",
        "ref": [
          "const"
        ],
        "rows_examined_per_scan": 766,
        "rows_produced_per_join": 255,
        "filtered": "33.33",
        "index_condition": "((`keywords`.`searches__request_dispatcher__msgs`.`time_acked` is null) and (`keywords`.`searches__request_dispatcher__msgs`.`time_next` < <cache>((unix_timestamp(now()) * 1e9))))",
        "cost_info": {
          "read_cost": "191.50",
          "eval_cost": "25.53",
          "prefix_cost": "268.10",
          "data_read_per_join": "181K"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ]
      }
    }
  }
}
```

Reply:
Continuing my non-scientific benchmark, I changed the where clause to max bigint, so it would now match all 3M rows, to give a better idea for the polling query on a backed up queue. This time, it did choose the new index by itself, and I had to force the `time_next` index. The new index is a clear winner here. I'm happy to test other permutations as well.

#### Test results:
- `time_next`: 350-400 ms
- `time_acked__priority__time_next`: 20-30 ms

```sql
select priority, time_next, epoch, time_acked, id, keyspace_id 
from searches__extractor__msgs # optional force index (time_next)
where time_acked is null and time_next < 9223372036854775807 
order by priority, time_next desc limit 10000
```


#### time_next query plan

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "121413.86"
    },
    "ordering_operation": {
      "using_filesort": true,
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "range",
        "possible_keys": [
          "time_next"
        ],
        "key": "time_next",
        "used_key_parts": [
          "time_next"
        ],
        "key_length": "9",
        "rows_examined_per_scan": 269808,
        "rows_produced_per_join": 26980,
        "filtered": "10.00",
        "index_condition": "(`keywords`.`searches__extractor__msgs`.`time_next` < 9223372036854775807)",
        "cost_info": {
          "read_cost": "118715.78",
          "eval_cost": "2698.08",
          "prefix_cost": "121413.86",
          "data_read_per_join": "18M"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ],
        "attached_condition": "(`keywords`.`searches__extractor__msgs`.`time_acked` is null)"
      }
    }
  }
}
```

#### time_acked__priority__time_next query plan

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "34230.00"
    },
    "ordering_operation": {
      "using_filesort": false,
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "ref",
        "possible_keys": [
          "time_acked__priority__time_next"
        ],
        "key": "time_acked__priority__time_next",
        "used_key_parts": [
          "time_acked"
        ],
        "key_length": "9",
        "ref": [
          "const"
        ],
        "rows_examined_per_scan": 267105,
        "rows_produced_per_join": 89026,
        "filtered": "33.33",
        "index_condition": "((`keywords`.`searches__extractor__msgs`.`time_acked` is null) and (`keywords`.`searches__extractor__msgs`.`time_next` < 9223372036854775807))",
        "cost_info": {
          "read_cost": "7519.50",
          "eval_cost": "8902.61",
          "prefix_cost": "34230.00",
          "data_read_per_join": "61M"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ]
      }
    }
  }
}
```

Reply:
I tried adding fields to your suggestion to make it a covering index, to see if that would change performance. It did speed it up 2-4x, but probably not worth it.

Test results:
- time_next: 350-400 ms
- time_acked__priority__time_next: 20-30 ms
- time_acked__priority__time_next__epoch__id__keyspace_id: 5-15 ms

```sql
alter table searches__extractor__msgs 
add index `time_acked__priority__time_next__epoch__id__keyspace_id` 
          (time_acked, priority, time_next desc, epoch, id, keyspace_id)
```

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "32684.49"
    },
    "ordering_operation": {
      "using_filesort": false,
      "table": {
        "table_name": "searches__extractor__msgs",
        "access_type": "ref",
        "possible_keys": [
          "ack_idx",
          "time_next",
          "time_acked__priority__time_next",
          "time_acked__priority__time_next__epoch__id__keyspace_id"
        ],
        "key": "time_acked__priority__time_next__epoch__id__keyspace_id",
        "used_key_parts": [
          "time_acked"
        ],
        "key_length": "9",
        "ref": [
          "const"
        ],
        "rows_examined_per_scan": 313005,
        "rows_produced_per_join": 156502,
        "filtered": "50.00",
        "using_index": true,
        "cost_info": {
          "read_cost": "1383.99",
          "eval_cost": "15650.25",
          "prefix_cost": "32684.49",
          "data_read_per_join": "108M"
        },
        "used_columns": [
          "id",
          "keyspace_id",
          "priority",
          "epoch",
          "time_next",
          "time_acked"
        ],
        "attached_condition": "((`keywords`.`searches__extractor__msgs`.`time_acked` is null) and (`keywords`.`searches__extractor__msgs`.`time_next` < 9223372036854775807))"
      }
    }
  }
}
```

Reply:
> Can you please include the query along with the explain plan? Can you also please share the index definitions? From your explain output here, I think you're using the new updated/proposed query but the new time_acked__priority__time_next index is not in DESC order for the time_next field.

Just saw this. I am using the updated query referencing `time_acked IS NULL`, and my multi-column indexes have always had `time_next desc`. There is a separate `time_next ASC` index that is being selected sometimes, since it matches the `time_next < ?` clause better. Does that answer your question?

```sql
  KEY `time_next` (`time_next`),
  KEY `time_acked__priority__time_next` (`time_acked`,`priority`,`time_next` DESC),
  KEY `time_acked__priority__time_next__epoch__id__keyspace_id` (`time_acked`,`priority`,`time_next` DESC,`epoch`,`id`,`keyspace_id`),
```

Reply:
I believe that I've now addressed the deadlocks. I am now leaning toward removing the query perf aspects and we can investigate and play with that more separately. What do you think? I would remove that index change from the docs PR too.

Reply:
Let's get it merged. I'm happy leaving the query perf in, as I did see some clear wins on the larger queries, but also happy to kick it to a separate PR

Reply:
OK, thanks! Sorry, I hadn't yet caught up on all of your more recent tests and comments. THANK YOU for that. I'll keep those changes in then.